### ISSUE TITLE
False-positive E1101 (no-member) for writable file-like object

### ISSUE THREAD

**Issue by tucked** [PyCQA/pylint#3234]
### Steps to reproduce
``` python
"""sscce.py"""

class Context:
    def __enter__(self):
        return self
    def __exit__(self, type_, exc, traceback):
        return False

class Iterable(Context):
    def __iter__(self):
        return self
    def __next__(self):
        return self
    def write(self, x):
        print(x)

class Foo(Iterable):
    pass

with Foo() as foo:
    foo.write('something')
```

Note: Like #3233, this was found while using `paramiko`:
``` python
        with self._sftp.open(local_path, mode) as local_file:
            local_file.write(contents)  # triggers on write
```
`local_file` is a `paramiko.SFTPFile` which inherits from `paramiko.BufferedFile` which inherits `paramiko.util.ClosingContextManager`.

### Current behavior
```
sscce.py:21:4: E1101: Instance of 'Context' has no 'write' member (no-member)
```

### Expected behavior
(no error)

### pylint --version output
```
pylint 2.4.3
astroid 2.3.3
Python 3.6.8 (default, Oct  7 2019, 12:59:55)
[GCC 8.3.0]
```

**Comment by simonvanderveldt** [comment]
Can confirm, just ran into this as well in our case we're using `paramiko.Transport` like this
```
with paramiko.Transport(...) as transport:
        transport.connect(...)
```

And we get this linting error
```
app/lambda_function.py:98:8: E1101: Instance of 'ClosingContextManager' has no 'connect' member (no-member)
```

I'm not even sure how it gets to the `ClosingContextManager` since the `Transport` class itself has a `connect()` member
https://github.com/paramiko/paramiko/blob/2.6.0/paramiko/transport.py#L1222

**Comment by PCManticore** [comment]
Thanks for the example, folks, I can confirm the bug.

**Comment by actionless** [comment]
i wonder, if it's the same issue or not:

```python
from typing import Any, List, Optional


NOT_FOUND_ATOM = object()


class AnnotatedDataType():

    def _key_not_exists(self, key):
        return getattr(self, key, NOT_FOUND_ATOM) is NOT_FOUND_ATOM

    def __init__(self, **kwargs) -> None:
        for key, value in kwargs.items():
            setattr(self, key, value)
        for key in self.__annotations__:  # pylint: disable=no-member
            if self._key_not_exists(key):
                raise TypeError(
                    f"'{self.__class__.__name__}' does "
                    f"not have required attribute '{key}' set"
                )

    def __setattr__(self, key: str, value: Any) -> None:
        if (
                not getattr(self, "__annotations__", None) or
                self.__annotations__.get(key, NOT_FOUND_ATOM) is NOT_FOUND_ATOM  # pylint: disable=no-member
        ) and self._key_not_exists(key):
            raise TypeError(
                f"'{self.__class__.__name__}' does "
                f"not have attribute '{key}'"
            )
        super().__setattr__(key, value)




class PackagesNotFound(AnnotatedDataType, Exception):
    packages: List[str]
    wanted_by: Optional[List[str]] = None

    def __init__(self, packages: List[str], wanted_by: Optional[List[str]] = None) -> None:
        AnnotatedDataType.__init__(self, packages=packages, wanted_by=wanted_by)
        message = ', '.join(packages)
        if wanted_by:
            message += f" wanted by {', '.join(wanted_by)}"
        Exception.__init__(self, message)


class PackagesNotFoundInRepo(PackagesNotFound):
    pass


def main():
    try:
        raise PackagesNotFoundInRepo(packages=['a', 'b', 'c'])
    except PackagesNotFound as exc:
        print(exc.packages)
    try:
        raise PackagesNotFoundInRepo(packages=['a', 'b', 'c'])
    except PackagesNotFoundInRepo as exc:
        print(exc.packages)
        #     ^ [pylint] no-member: Instance of 'PackagesNotFoundInRepo' has no 'packages' member


if __name__ == "__main__":
    main()
```

**Comment by actionless** [comment]
i slightly updated my previous message, because the main point was not the annotations (which also indeed are not working there)

**Comment by fenuks** [comment]
As for `__annotations__`, `no-member` warning is indeed also issued for other kind of typed things in `typing` module, e.g. `NamedTuple`:

```python
from typing import NamedTuple


class Foo(NamedTuple):
    foobar: str

print(Foo.__annotations__)
```
```
test.py:7:6: E1101: Class 'Foo' has no '__annotations__' member (no-member)
```

```sh
$ pylint --version
pylint 2.10.2
astroid 2.7.2
Python 3.9.6 (default, Aug 23 2021, 11:20:01) 
 ```

**Comment by cdce8p** [comment]
This issue seems to be fixed when testing with the latest pylint version.
If you still encounter issues, please report back and I'll reopen it.

**Comment by fenuks** [comment]
Latest tagged version? With 
```
$ pylint --version
pylint 2.11.1
astroid 2.8.2
Python 3.9.7 (default, Oct 10 2021, 15:13:22) 
[GCC 11.1.0]
```
for the example I posted above problem still persists.

**Comment by cdce8p** [comment]
@fenuks I'd tested it with Python 3.10 which works.
However, you're right at least `__annotations__` is still broken for Python < 3.10

I can't reproduce the original issue with `Context.write` though. Even not with Python 3.8

**Comment by Pierre-Sassoulas** [comment]
Closing as the main issue was fixed, and I opened https://github.com/PyCQA/pylint/issues/7126 for the one that appeared along the way.
